Fix the prompt to install jetpack not shown when logout WP account

ipalm0423 opened the pull request:
Fixes #19398

| -- | before | after |
| -- | -- | -- |
| iPhone | ![Simulator Screen Shot - iPhone 14 - 2022-10-06 at 02 53 18](https://user-images.githubusercontent.com/10288494/194139502-d9865a14-a0b5-4bbc-b31f-f5ae6b33a329.png) | ![Simulator Screen Shot - iPhone 14 - 2022-10-06 at 02 10 36](https://user-images.githubusercontent.com/10288494/194137937-990785df-1c6a-43f5-902c-4749ae04cff2.png) |
| iPad | ![Simulator Screen Shot - iPad Pro (9 7-inch) - 2022-10-06 at 02 49 06](https://user-images.githubusercontent.com/10288494/194138790-01d4f39e-4341-4bff-a82a-928f398e4074.png) | ![Simulator Screen Shot - iPad Pro (9 7-inch) - 2022-10-06 at 02 13 56](https://user-images.githubusercontent.com/10288494/194137988-c9970ffa-8c2f-416c-b346-bdf2647f55b1.png) |

## Root Cause
When the user logout the WP.com account. It will trigger `MySiteViewController` [showBlogDetailsForMainBlogOrNoSites()](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Blog/My%20Site/MySiteViewController.swift#L408) to display an available blog on the page. But it doesn't update the `RecentSitesService` for blog usage history.

So, when the `Notifications` tab appear, it couldn't get the correct [last-used](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController%2BJetpackPrompt.swift#L4) blog for displaying jetpack prompt.

## Fix
801e14bb08720a697c0615454b06ec5579af9394 :
- update `RecentSitesService`'s history when the `MySiteViewController` change it's blog.

ea0339c3b25523d90aa6f5a1af4b3d0c2d9cdee2 :
- At `showBlogDetailsForMainBlogOrNoSites` method, and it should call `createFABIfNeeded`, `fetchPrompt(for: newBlog)` when changing blog. So remove the duplicate code by using `didSet` in `self.blog`

04a9c84e4010afa3a3bddbcd968bb5333e41c0da 
aff4383ab2a1ec311b8ed7d6320a5f2125d34d68 :
- fix Jetpack banner hidden logic when scrolling
- update jetpack banner appearance when `Notifications` tab `viewDidLoad`, `defaultAccountDidChange`

97bb5d50c4a664758d05cf29f07d227640a21d48 : 
- Fix empty view not in the center if the tableView has been scroll before logout.

| before | after |
| -- | -- |
| ![Simulator Screen Shot - iPhone 14 - 2022-10-08 at 02 40 17](https://user-images.githubusercontent.com/10288494/194640969-2a5550f7-cedb-4b4b-b8ac-fc138f0b9faa.png) | ![Simulator Screen Shot - iPhone 14 - 2022-10-06 at 02 10 36](https://user-images.githubusercontent.com/10288494/194640905-6fc8fbab-0464-42f5-a15f-8795cbc5ca0d.png) |


## To test:
1. With a clean install of the WordPress app, log into a self-hosted site that doesn't have Jetpack connected
2. Tap the "Notifications" tab
3. Observe the button to install Jetpack and the text "To get helpful notifications on your phone from your WordPress site..."
4. Tap "My Site" and present the "Me" view by tapping the avatar at the top right
5. Log into a WordPress.com account
6. Tap the "Notifications" tab
7. Observe a list of notifications and the "Jetpack powered" banner at the bottom (if no notifications exist for the account, you'll see a "No notifications yet" message)
8. Return to "My Site" -> "Me" and log out of the WordPress.com account
9. Tap the "Notifications" tab
10. observe jetpack prompt and without the "Jetpack powered" banner at the bottom

## Related Issue:
Currently, there is a bug in the onboarding login process. ([issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/19419)). 
If you select `skip` on the onboarding page `What would you like to focus on first?`.
The onboarding process will cause a wrong history of `RecentSitesService` and let the `Notifications` tab won't change to the `Install Jetpack` view after you logout,. 
It's a known issue and will fix it later with the [issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/19419).

## Regression Notes
1. Potential unintended areas of impact
- The jetpack banner on the `Notifications` tab
- switch account behavior on `My Site` tab

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing on [iPhone/iPad] with [WordPress.com/self-hosted] account switching.

3. What automated tests I added (or what prevented me from doing so)
NA, This is a visual fix.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ipalm0423 committed:
touch `RecentSitesService` when change blog in `MySiteViewController

ipalm0423 commented:
When tested on `trunk`, I found another weird behavior that might be a bug🤔
Please let me know If it's a bug, then I will open a new issue to fix it:

Reproduce steps:
1. Login the `self-hosted` site 
2. Login the `WordPress.com` account. 
3. On the onboarding page select a blog to open.
4. After onBoarding, it doesn't change to the blog I select on the `My site` tab. It keeps on `self-hosted` site.


| - | onBoarding select a blog on list | select `Checking Stats` | after onBoarding `My Site` tab |
| -- | -- | -- | -- |
| -- | ![masz](https://user-images.githubusercontent.com/10288494/194645910-49ad0346-6622-4c37-aa48-f8bd91c43f89.png) | ![Simulator Screen Shot - iPhone 14 - 2022-10-06 at 03 30 58](https://user-images.githubusercontent.com/10288494/194147641-5bca2333-fc47-4f0b-8f90-623d1eeafea4.png) | ![Simulator Screen Shot - iPhone 14 - 2022-10-06 at 03 31 08](https://user-images.githubusercontent.com/10288494/194147719-ef8654a6-841b-4b9d-a1ae-f9fa1d757b90.png) |

twstokes commented:
CI PR created: https://github.com/wordpress-mobile/WordPress-iOS/pull/19408

twstokes commented on RELEASE-NOTES.txt:
```diff
@@ -1,6 +1,7 @@
 21.0
 -----
 * [*] Fixed an issue where the cached notifications are retained after logging out of WordPress.com account [#19360]
+* [*] Fixed an issue where the prompt to install jetpack not shown when logout WorldPress.com account [#19403]
```
```suggestion
* [*] Fixed an issue where the prompt to install Jetpack is not shown when logout WorldPress.com account [#19403]
```

twstokes replied:
Nit: A minor spelling + grammar tweak.

ipalm0423 replied:
ok, modified at 97bb5d50c4a664758d05cf29f07d227640a21d48

twstokes reviewed (changes requested):
👋 Thanks for working on this @ipalm0423!

I was still able to get the Jetpack banner to appear after removing the WordPress.com account and only using a self-hosted site without a Jetpack connection:


https://user-images.githubusercontent.com/2092798/194336324-5b789eec-6e0b-4821-aef6-e599f1a5e8d6.mp4

ipalm0423 commented:
> 👋 Thanks for working on this @ipalm0423!
> 
> I was still able to get the Jetpack banner to appear after removing the WordPress.com account and only using a self-hosted site without a Jetpack connection:
> 
>  Banner.visible.mp4

Thank you for the review. 🙏 @twstokes 
The scrolling gesture causes this bug.
Fix it on aff4383ab2a1ec311b8ed7d6320a5f2125d34d68
and test on different page for jetpack banner.

| Reader search | Reader tab | Stats | Activity log | Period |
| -- | -- | -- | -- | -- |
| ![Simulator Screen Shot - iPhone 14 - 2022-10-08 at 05 10 27](https://user-images.githubusercontent.com/10288494/194653425-2546fd40-e656-4ee0-b190-c3d016da743c.png) | ![Simulator Screen Shot - iPhone 14 - 2022-10-08 at 02 07 27](https://user-images.githubusercontent.com/10288494/194653470-f047d7b9-dc1c-49ba-a010-3bbab0404bc4.png) | ![Simulator Screen Shot - iPhone 14 - 2022-10-08 at 02 06 20](https://user-images.githubusercontent.com/10288494/194653530-1ec89d8b-52f9-48c4-a68f-9150c0d784eb.png) | ![Simulator Screen Shot - iPhone 14 - 2022-10-08 at 02 06 49](https://user-images.githubusercontent.com/10288494/194653607-ab036ad0-1c6d-42b0-b666-9317185e9490.png) | ![Simulator Screen Shot - iPhone 14 - 2022-10-08 at 05 23 03](https://user-images.githubusercontent.com/10288494/194656089-1ab7f7e4-d679-47f4-8bb2-5649a2cecc29.png) |

Also, I rebased the branch to `trunk` and fixed more bugs on the `Notifications` tab with the new commits.

ipalm0423 commented:
> When tested on `trunk`, I found another weird behavior that might be a bug🤔 Please let me know If it's a bug, then I will open a new issue to fix it:
> 
> Reproduce steps:
> 
> 1. Login the `self-hosted` site
> 2. Login the `WordPress.com` account.
> 3. On the onboarding page select a blog to open.
> 4. After onBoarding, it doesn't change to the blog I select on the `My site` tab. It keeps on `self-hosted` site.
> 

After searching the existing issues on GitHub, I only found a similar [one](https://github.com/wordpress-mobile/WordPress-iOS/issues/7455).  But it's not the same as this.
So I open a new [one](https://github.com/wordpress-mobile/WordPress-iOS/issues/19419).

twstokes commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:
```diff
@@ -627,18 +627,17 @@ private extension NotificationsViewController {
-        jetpackBannerView.buttonAction = { [unowned self] in
-            JetpackBrandingCoordinator.presentOverlay(from: self)
-            JetpackBrandingAnalyticsHelper.trackJetpackPoweredBannerTapped(screen: .notifications)
+        if jetpackBannerView.buttonAction == nil {
```
👋 @ipalm0423 - Could we keep the logic to determine whether or not to activate the banner local to this view?

We've followed this pattern for the other views, and although they currently all match the same logic, we aren't sure if some may differ in the future. Thanks!

ipalm0423 replied:
👋 @twstokes Ok, No problem. 
I change my fix: 
aff4383ab2a1ec311b8ed7d6320a5f2125d34d68 
04a9c84e4010afa3a3bddbcd968bb5333e41c0da
- These commits only impacts the `NotificationsViewController`, which makes the impaction smaller and easy to test. After we know every specification on other views, we can open another PR to fix it.
- Change the Jetpack view appearance only when `viewDidLoad`/`accountChanged`.

ipalm0423 committed:
reload JetpackBanner if defaultAccountDidChange

ipalm0423 committed:
processJetpackBannerVisibility only when JetpackBrandingVisibility is enabled

ipalm0423 committed:
reset contentOffset when display placeholder on tableView

twstokes commented on RELEASE-NOTES.txt:
```diff
@@ -10,6 +10,7 @@
 * [*] [internal] A refactor in weekly roundup notification scheduler. [#19422]
 * [*] Stories: Fixed an issue where the keyboard would overlap with the publish dialog in landscape. [#19350]
 * [*] Fixed an issue where the navigation bar becomes invisible when swiping back to Login Prologue screen.  [#19461]
+* [*] Fixed an issue where the prompt to install Jetpack is not shown when logout WordPress.com account [#19403]
```
@ipalm0423 do you think we can further improve this release note?

I'm thinking from the perspective of someone who doesn't have much context on the issue - could we:
- sum up this specific issue better ("Jetpack is not shown when logout WordPress.com account" may not make sense to the editorial team)
- match previous release note formatting + grammar

Thanks!

ipalm0423 replied:
How about: `Fixed an issue where the notification tab's prompt view doesn't match the login account`

I want to keep the note as short as possible because it already contains the issue number `[#19403]`. If someone is interested in the fix, they can open the issue easily.

twstokes commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift:
```diff
@@ -410,11 +410,7 @@ class MySiteViewController: UIViewController, NoResultsViewHost {
-        showBlogDetails(for: mainBlog)
-        updateNavigationTitle(for: mainBlog)
-        updateSegmentedControl(for: mainBlog, switchTabsIfNeeded: true)
+        self.blog = mainBlog
```
This seems like the right direction towards a more DRY approach.

- Do you foresee any issues with the extra calls to `createFABIfNeeded()` and `fetchPrompt(for: newBlog)` when this is set?

ipalm0423 replied:
> This seems like the right direction towards a more DRY approach.

Yes, It's easy to maintain if we toward DRY in code. 

> - Do you foresee any issues with the extra calls to createFABIfNeeded() and fetchPrompt(for: newBlog) when this is set?

About the extra call, it will have minor effect on the `createFABIfNeeded()`:
- Inside the `createFABIfNeeded` it increase the counting for the `CreateButtonCoordinator.shownTooltipCount`.
If we has extra cell on it, it will have extra counting for the tool tips.

Because it have the side effect above if we only tweak code here, 
It will be better to refactor for DRY all at once in the future, so I revert the tweak for this PR.

twstokes commented:
> > When tested on `trunk`, I found another weird behavior that might be a bug🤔 Please let me know If it's a bug, then I will open a new issue to fix it:
> > Reproduce steps:
> > 
> > 1. Login the `self-hosted` site
> > 2. Login the `WordPress.com` account.
> > 3. On the onboarding page select a blog to open.
> > 4. After onBoarding, it doesn't change to the blog I select on the `My site` tab. It keeps on `self-hosted` site.
> 
> After searching the existing issues on GitHub, I only found a similar [one](https://github.com/wordpress-mobile/WordPress-iOS/issues/7455). But it's not the same as this. So I open a new [one](https://github.com/wordpress-mobile/WordPress-iOS/issues/19419).

Thanks! Looks like a bug to me.

twstokes reviewed (commented):
👋 Hi @ipalm0423 -

When I:
1. Add a self-hosted site that doesn't have Jetpack installed
2. Log into a WordPress.com account that has notifications
3. Log out of the WordPress.com account
4. Go to the Notifications view

I see the "No notifications yet" message when I expect to see "Install Jetpack". This does not occur for me if the WordPress.com account had zero notifications.

<img src="https://user-images.githubusercontent.com/2092798/196602624-58f3507f-5d0e-437c-8109-d4e908504e9f.png" width="350px" />

ipalm0423 commented:
> 👋 Hi @ipalm0423 -
> 
> When I:
> 
> 1. Add a self-hosted site that doesn't have Jetpack installed
> 2. Log into a WordPress.com account that has notifications
> 3. Log out of the WordPress.com account
> 4. Go to the Notifications view
> 
> I see the "No notifications yet" message when I expect to see "Install Jetpack". This does not occur for me if the WordPress.com account had zero notifications.
> 
👋  Hi @twstokes :
I believe it's a bug coming from the [issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/19419) mentioned above:
In the step `2` , when you choose site on the onboarding page, sometimes it doesn't change the site in `My Site` tab, but it does change the history on `RecentSitesService`. 
This will cause a mismatch between the history and the `My Site` tab, and the mismatch keeps wrong when you logout of the WP.com account and causes this bug.
<img width="248" alt="Screen Shot 2022-10-22 at 1 20 07 AM" src="https://user-images.githubusercontent.com/10288494/197253062-d24ff317-a009-4d32-90f5-87523152f476.png">

###  Reproduce Steps with detail:
> 1. Add a self-hosted site that doesn't have Jetpack installed

| on `My Sites` tab | in `RecentSitesService` |
| - | - |
| self-hosted site | self-hosted site |

> 2. Log into a WordPress.com account that has notifications

| on `My Sites` tab | in `RecentSitesService` |
| - | - |
| self-hosted site | the WP site you selected on the onboarding page |

**History Mismatch:** the `My Sites` tab doesn't replace to the WP site you selected on the onboarding page.

> 3. Log out of the WordPress.com account

On `My Sites` tab, because it use the self-hosted site already, so it won't trigger the site update flow in [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Blog/My%20Site/MySiteViewController.swift#L408), so the `RecentSitesService` history keep wrong.

| on `My Sites` tab | in `RecentSitesService` |
| - | - |
| self-hosted site | a WP site you select on the onboarding page |

> 4. Go to the Notifications view

The `NotificationsViewController` read the wrong history on `RecentSitesService` and present a wrong prompt view.


### Conclusion:
The wrong prompt view is caused by the wrong history on the `RecentSitesService`.
And there has two issues to make history wrong:
- Login part: update the history correctly but doesn't update the site on `My Site` tab ( [issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/19419))
- Logout part: redirect to self-hosted site without update the history (fix by this PR)

ipalm0423 committed:
Update draft release notes for 21.1

mokagio commented:
Bumping the milestone from 21.1 to 21.2 because this hasn't been approved yet and I'm running the 21.1 code freeze today.

If this cannot wait two weeks and it's important that it makes it into this release, let me know and we'll organize a new beta once ready.
